soc/nordic/common/dmm.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * nRF SoC specific public APIs for Device Memory Management (dmm) subsystem
+ */
+
+#ifndef SOC_NORDIC_COMMON_DMM_H_
+#define SOC_NORDIC_COMMON_DMM_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @cond INTERNAL_HIDDEN */
+
+/**
+ * @brief Prepare a DMA output buffer for the specified device
+ *
+ * Allocate an output buffer in memory region that given device can perform DMA transfers from.
+ * Copy @p user_buffer contents into it.
+ * Writeback data cache lines associated with output buffer, if needed.
+ *
+ * @note Depending on provided user buffer parameters and SoC architecture,
+ *       dynamic allocation and cache operations might be skipped.
+ *
+ * @warning It is prohibited to read or write @p user_buffer or @p buffer_out contents
+ *          until @ref dmm_dma_buffer_out_release() is called
+ *
+ * @param dev Device to prepare the buffer for.
+ * @param user_buffer Buffer containing data to be processed by the given device
+ * @param user_out Pointer to a pointer to be filled by prepared DMA buffer address.
+ *
+ * @retval 0 If succeeded.
+ * @retval -ENOMEM If output buffer could not be allocated.
+ * @retval -errno Negative errno for other failures.
+ */
+int dmm_dma_buffer_out_prepare(const struct device *dev, void const * user_buffer, void ** buffer_out);
+
+/**
+ * @brief Release the previously prepared DMA output buffer
+ *
+ * @param dev Device to release the buffer for.
+ * @param buffer_out DMA output buffer previously prepared with @ref dmm_dma_buffer_out_prepare().
+ *
+ * @retval 0 If succeeded.
+ * @retval -errno Negative errno code on failure.
+ */
+int dmm_dma_buffer_out_release(const struct device *dev, void const * buffer_out);
+
+/**
+ * @brief Prepare a DMA input buffer for the specified device
+ *
+ * Allocate an input buffer in memory region that given device can perform DMA transfers to.
+ *
+ * @note Depending on provided user buffer parameters and SoC architecture,
+ *       dynamic allocation might be skipped.
+ *
+ * @warning It is prohibited to read or write @p user_buffer or @p buffer_in contents
+ *          until @ref dmm_dma_buffer_in_release() is called
+ *
+ * @param dev Device to prepare the buffer for.
+ * @param user_buffer Buffer to be filled with data from the given device
+ * @param user_in Pointer to a pointer to be filled by prepared DMA buffer address.
+ *
+ * @retval 0 If succeeded.
+ * @retval -ENOMEM If input buffer could not be allocated.
+ * @retval -errno Negative errno for other failures.
+ */
+int dmm_dma_buffer_in_prepare(const struct device *dev, void const * user_buffer, void ** buffer_in);
```

Comment:
I wonder if this function should not be split into `alloc` and `prepare`. Currently this api does not cover case when driver would like to allocated at compile time a buffer (e.g UART for polling out or for RX flushing) and then use this API only for cache management (if we want to use only dmm for cache management). `alloc` might be common for both directions.

For dynamic buffer it would look like
```
err = dmm_dma_buffer_alloc(dev, &dma_buffer, length);
if (err == 0) dmm_dma_buffer_out_prepare(dev, dma_buffer, user_buffer, length);
```
For statically allocated buffers
```
uint8_t dma_buffer[length] MEM_REGION(...);

// Prepare function just copies and performs cache writeback (if needed)
dmm_dma_buffer_out_prepare(dev, dma_buffer, user_buffer, length2); //length2 <= length
```

Reply:
As a next step dmm API should be extended with macros for static buffer allocation. These macros can be used by the drivers covering the scenario you described, or by the applications to improve the performance of the drivers.

If the buffers are allocated using the to-be-created dmm macros, they are guaranteed to be bypassed by `dmm_dma_buffer_in/out_prepare()` without dynamic buffer allocation and copying overhead.

I don't think the driver should have any indication if the buffer in use is to be bypassed or not by dmm. The driver should not track the qualities of the buffers, because it's dmm's responsibility. The driver should blindly call `dmm_..._prepare()` on any buffer.

Reply:
> Currently this api does not cover case when driver would like to allocated at compile time a buffer (e.g UART for polling out or for RX flushing) and then use this API only for cache management (if we want to use only dmm for cache management). 

Actually this API covers that - if the buffer has correct attributes (location and cache alignment) then no dynamic allocation will happen and `dmm` will simply return what was provided to it